### PR TITLE
Allow registering external IDPs.

### PR DESCRIPTION
**Pull request by VirajSalaka**
Objective is to support third party IDPs in loosely coupled manner from api manager point of view. API manager will not communicate with the IDP at all, but it would store it so that the gateway can directly call the external IDP and validate all by itself. 

We bring in the tokenType call `EXTERNAL` to support this feature. And when for this type, issuer, JWKS URL, Token Endpoint, Authorize endpoint, Revoke Endpoint, Scope claim key and ConsumerKey claim key is mandatory.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml [hunk 1]
```diff
@@ -1218,6 +1218,7 @@ definitions:
           - EXCHANGED
           - DIRECT
           - BOTH
+          - EXTERNAL
```
Comment (VirajSalaka) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/resources/api.yaml:1221]: Remove BOTH type from the Rest API level

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java [hunk 2]
```diff
@@ -725,14 +776,39 @@ private String getDecryptedValue(String value) throws APIManagementException {
...
+        if (oldKeyManagerConfiguration == null) {
+            throw new APIManagementException("Key Manager Configuration not found for ID: "
+                    + keyManagerConfigurationDTO.getUuid(), ExceptionCodes.KEY_MANAGER_NOT_FOUND);
+        }
+        if (KeyManagerConfiguration.TokenType.EXTERNAL.equals(keyManagerTokenType)
+                ^ KeyManagerConfiguration.TokenType.EXTERNAL.equals(
+                        KeyManagerConfiguration.TokenType.valueOf(oldKeyManagerConfiguration.getTokenType()
+                                .toUpperCase()))) {
+            throw new APIManagementException("Cannot switch between tokenType when " +
+                    "the Key Manager's Token Type is External", ExceptionCodes.KEY_MANAGER_NOT_FOUND);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java:801]: Let's fix the Error code

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java [hunk 3]
```diff
@@ -116,6 +118,20 @@ public class APIAdminImpl implements APIAdmin {
...
+            APIConstants.KeyManager.SCOPES_CLAIM,
+            APIConstants.KeyManager.CERTIFICATE_TYPE,
+            APIConstants.KeyManager.CERTIFICATE_VALUE
+    };
+
+    private static final String[] requiredEndpointConfigurationsForExternalKeyManagers = {
+            APIConstants.KeyManager.AUTHORIZE_ENDPOINT,
+            APIConstants.KeyManager.TOKEN_ENDPOINT,
+            APIConstants.KeyManager.REVOKE_ENDPOINT
+    };
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java:133]: Shall we move this to APIConstants or any other better place

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java [hunk 4]
```diff
@@ -116,6 +118,20 @@ public class APIAdminImpl implements APIAdmin {
     protected ApiMgtDAO apiMgtDAO;
     protected ChoreoApiMgtDAO choreoApiMgtDAO;
 
+    private static final String[] requiredConfigurationsForExternalKeyManagers = {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIAdminImpl.java:121]:
```suggestion
    private static final String[] REQUIRED_CONF_FOR_EXT_KMS = {
```

**Review by malinthaprasan — APPROVED** [review]

**Review by malinthaprasan — APPROVED** [review]

**Pull request merged**